### ISSUE TITLE
'None' Optional models are excluded from queries, including selects

### ISSUE THREAD

**Issue by xeals** [moigagoo/norm#79]
Using norm version 2.1.1.

Minimal reproduction:

```nim
import logging, options
import norm/[model, sqlite]

addHandler(newConsoleLogger())

type
  A = ref object of Model
    a: int
  B = ref object of Model
    a: Option[A]
    b: int

func newB(): B = B(a: none A, b: 0)

when isMainModule:
  withDb:
    db.createTables(newB())
```

I'd expect it to create tables for both `A` and `B` here, as `B` depends on `A`, but the output I get is only (formatted for readability):

```
DEBUG CREATE TABLE IF NOT EXISTS "B"(
  a INTEGER,
  b INTEGER NOT NULL,
  id INTEGER NOT NULL PRIMARY KEY,
  FOREIGN KEY(a) REFERENCES "A"(id)
)
```

The same behavior is observed with querying, updating, etc:

```nim
discard newB().dup:
  db.select("B.b > ?", 0)
```

results in not joining onto the `A` table:

```
DEBUG SELECT "B".b, "B".id FROM "B"  WHERE B.b > ? <- [0]
```

For comparison, changing `B.a` to be just an `A` and adjusting `newB()` accordingly results in the following output instead, matching the README:

```
DEBUG CREATE TABLE IF NOT EXISTS "A"(
  a INTEGER NOT NULL,
  id INTEGER NOT NULL PRIMARY KEY
)
DEBUG CREATE TABLE IF NOT EXISTS "B"(
  a INTEGER NOT NULL,
  b INTEGER NOT NULL,
  id INTEGER NOT NULL PRIMARY KEY,
  FOREIGN KEY(a) REFERENCES "A"(id)
)
DEBUG SELECT "A".a, "A".id, "B".b, "B".id FROM "B" JOIN "A" ON "B".a = "A".id WHERE B.b > ? <- [0]
```

**Comment by xeals** [comment]
I've done some reading and some playing and this isn't so much an issue as described in the OP but just presents a challenge with SQLite and null values. I'm not sure if this goes for Postgres as well.

The current implementation generates joins for `some` optional values, which is reflected above where all of my testing generates a `none` for the submodel. This makes sense when inserting and updating -- no need to make joins on tables you're not modifying.

What led me to report this and my actual issue with the implementation is with querying. The select query generates joins the same way inserting does, which has strong implications for how you structure your `newModel` proc due to how the database handle joins on null columns. Using the example in the OP:

- if `newB()` has `B.a = none A`, the joins are not generated and nothing related to `B.a` will be present in the query (no `B.a`, `A.id`, or the join). If any `B`s in the table have non-null `A`s, those nested models will not be returned.
- if `newA()` has `B.a = some newA()`, the joins are generated and `B.a` will be populated; however, because of the `JOIN "A" on "B".a = "A".id`, any `B`s that have null `A` will not be returned.

I'm not sure what the best way to solve this is, or if it can be solved in a single query. I'm not familiar enough with other ORMs to know how they've solved it.

**Comment by moigagoo** [comment]
@xeals Thanks for reporting the issue!

There are to separate issues, I'll address them one by one.

**> if newB() has B.a = none A, the joins are not generated and nothing related to B.a will be present in the query (no B.a, A.id, or the join). If any Bs in the table have non-null As, those nested models will not be returned.**

This is more of a design choice than an issue. You see, unlike Norm 1, Norm 2 always requires an instance to work with. Be it creating a table or inserting a row or selecting a row—you must give it an object instance to infer types from, read values from, and store the data to respectively.

In your example, no instance of `A` is created so Norm doesn't have a container to store the selected row to. So it doesn't select it. In this trivial example it may appear logical to instantiate `A` with `A()` and Norm 1 did exactly that, but this approach breaks really badly for object fields that don't have default values, e.g. `DateTime`.

So instead of trying to instantiate objects, Norm 2 requires the programmer to do that and pass a ready to use instance. If the programmer does not give `select` a container to store the data then the data is not fetched.

**> because of the JOIN "A" on "B".a = "A".id, any Bs that have null A will not be returned. This doesn't crop up joining tables using LEFT JOIN rather than regular inner JOIN, but I'm not sure of the performance penalties in using it (if any).**

This is a serious issue, thanks for reporting it. I can see another solution in using `ON "B".a = "A".id OR "B".a IS NULL` instead of just `ON "B".a = "A".id` but `LEFT JOIN` looks much better. The queries are cleaner and more importantly, the returned rows are valid. Compare:
```
# This is valid, one B.a is NULL and the other is not:
sqlite> SELECT "A".a, "A".id, "B".b, "B".id FROM "B" LEFT JOIN "A" ON "B".a = "A".id WHERE 1;
0|1|0|1
||123|2

# This is invalid, A.a and A.id are not NULL in both rows:
sqlite> SELECT "A".a, "A".id, "B".b, "B".id FROM "B" JOIN "A" ON "B".a = "A".id OR "B".a IS NULL WHERE 1;
0|1|0|1
0|1|123|2
```
I'm not sure whether the performance will suffer from using `LEFT JOIN` instead of `INNER JOIN` but I think this is the best solution. Even if it gets us a performance penalty, it's a fair price to pay for proper `select`.

**Comment by moigagoo** [comment]
Quick update. I'm working on the fix but it turns out that the fix is far from trivial. I'm rewriting the way select queries are generated as well as the way roes are converted to objects.

**Comment by moigagoo** [comment]
Nite to self:
1. id field must be of type RowId = distinct int64 instead of int
2. There must be a to func that takes a RowId and a seq[DbValue] and returns either a none or some Model.
3. There must be a to func that takes a RowId and a seq[DbValue] and returns a Model.
4. to funcs should operate on values instead of typedescs.

**Comment by moigagoo** [comment]
This is getting fixed in https://github.com/moigagoo/norm/pull/88

**Comment by moigagoo** [comment]
Fixed with #88
